ramips: add support for ZBT-WE1326-v5

renamed98324 opened the pull request:
ramips: add support for ZBT-WE1326-v5
ZBT-WE1326 hardware revision 5 has 256MiB RAM (Nanya NT5CC128M16IP-DI) instead of 512MiB.
Specification:
- SoC: MT7621AT, MT7603EN and MT7612EN
- Flash: 16 MiB (W25Q128FVSG)
- RAM: 256 MiB (NT5CC128M16IP-DI)
- Ethernet: 1 x WAN (10/100/1000Mbps) and 4 x LAN (10/100/1000 Mbps)
- Others: USB 3.0, micro SD slot, reset button and 8 x LEDs

How to install:
- The original firmware in v5 does not allow to upload OpenWrt image file, so the procedure is to upload from recovery.
- Follow this steps to access recovery and write firmware.
1 - Configure LAN interface on 192.168.1.0/24 subnet or enable DCHP (recovery mode has DHCP server)
2 - Press and hold reset button for 5-10 seconds while powering the device.
3 - Access 192.168.1.1 click on "恢复" and upload "squashfs-sysupgrade.bin" image file. DO NOT upload "squashfs-factory" image file, as it will not work correctly (because Openwrt will boot but filesystems are not correctly mounted).

Signed-off-by: David Bosque <contact@wifiseguro.es>

wifiseguro commented:
Hello,
I just compiled ZBT-WE1326-v5, and flashed the sysupgrade.bin file and cannot boot. To restore access to the device I had to compile ZBT-WE1326 after editing its dts file to set 256 MB of RAM instead of 512, and it works, flashed and I can access the device again.
What could I do?

wifiseguro commented:
Osrry, this is my first pull request and I am quite new at github too, I tried to edit comit commentary and fix some errors in first commit and maybe I messed it a  bit because I made 4 commits and a merge of branchs, unfortunately the old comentary is still showing.

ynezz commented:
> Sorry, this is my first pull request and I am quite new at github too, I tried to edit comit commentary and fix some errors in first commit and maybe I messed it a bit because I made 4 commits and a merge of branchs, unfortunately the old comentary is still showing.

Please don't close this PR and open a new one. Please use `git commit --amend` to add changes to your commit. Use `git push --force` to overwrite your github repository/branch. The PR will afterwards update automatically. See the [squashing commits instructions](https://openwrt.org/docs/guide-developer/working-with-github-pr#squashing_commits) on how to modify this PR without closing it and open a new one.

```
git reset --soft HEAD~2
git commit -a -c ORIG_HEAD
```

or

```
git rebase -i HEAD~2
# now select the last commit and change `pickup` to `squash`
```

Commit:
ramips: add support for ZBT-WE1326-v5
ZBT-WE1326 hardware revision 5 has 256MiB RAM (Nanya NT5CC128M16IP-DI) instead of 512MiB.
Specification:

    SoC: MT7621AT, MT7603EN and MT7612EN
    Flash: 16 MiB (W25Q128FVSG)
    RAM: 256 MiB (NT5CC128M16IP-DI)
    Ethernet: 1 x WAN (10/100/1000Mbps) and 4 x LAN (10/100/1000 Mbps)
    Others: USB 3.0, micro SD slot, reset button and 8 x LEDs

How to install:

    The original firmware in v5 does not allow to upload OpenWrt image file, so the procedure is to upload from recovery.
    Follow this steps to access recovery and write firmware.
    1 - Configure LAN interface on 192.168.1.0/24 subnet or enable DCHP (recovery mode has DHCP server)
    2 - Press and hold reset button for 5-10 seconds while powering on the device.
    3 - Access 192.168.1.1 click on "恢复" and upload "squashfs-sysupgrade.bin" image file. DO NOT upload "squashfs-factory.bin" image file, as it will not work correctly (because Openwrt will boot but filesystems are not correctly mounted).

wifiseguro commented:
Hello Ynezz,
Thank you for the instructions, I followed them and now we it is fixed, one commit with the 3 files modified to support the new hardware version of ZBT-WE1326-v5.
Please let me know if anything else is needed.

adrianschmutzler commented:
There has been a reorganization in ramips target today. Please:
- update your board.d subfiles since they have been re-sorted
- update your Makefile (mt****/rt****.mk) changes so sorting of devices is correct
- rename your DTS/DTSI files following the scheme soc_vendor_device.dts(i), e.g. mt7620a_tplink_archer-c50-v1.dts (and update includes there if necessary)
- remove `DTS := ...` from Makefile
- add `MTK_SOC` to Makefile (needs to match "soc" in DTS file name), e.g. `MTK_SOC := mt7620a`

Thank you!

Also note that due to another recent change you will have to split DEVICE_TITLE into DEVICE_VENDOR, DEVICE_MODEL and DEVICE_VARIANT.

ynezz commented:
>  I followed them and now it is fixed

No it's not.

wifiseguro commented:
@ynezz  thank you for the follow up, I will check it again, and also will check ramips reorganization as per adrianschmutzler comment.

adrianschmutzler commented:
@wifiseguro 
To me it looks like this device is similar to the existing WE1326 except for the memory size.
Since memory is auto-detected in recent master, anything else would be equivalent.

In other words: You should be able to use the current existing snapshot for WE1326 already.

So, to me this looks like the PR is actually obsolete, and you can save the time to update it.
Feel free to correct me if I'm wrong.

adrianschmutzler commented:
For reference:
https://github.com/openwrt/openwrt/commit/6d91ddf5175d2eac3c4bc4a404cc0f5dd44cf25b
https://github.com/openwrt/openwrt/commit/a2c19f1d2f658367e6d62a6bdcfc72f12f23e43e

wifiseguro commented:
@adrianschmutzler 
I thought the same when when I red about the memory auto-detect update for mt7621, about one month ago. Nice get confirmation from you.

adrianschmutzler commented:
Yes, would be nice if you could just test with the latest snapshot:
https://downloads.openwrt.org/snapshots/targets/ramips/mt7621/

If that works, I'd say we do not need to introduce a separate v5 and you can just close the PR.
If it doesn't work for you, we'll have to investigate.

wifiseguro commented:
@adrianschmutzler

I cloned master branch yesterday and compiled a  firmware, and I confirm that ir works as expected, the firmware detectes 256 MB of Ram and booted with no issues, so definetly there is no need for this push request anymore.
Thank you for you kind support.
Should you need anything else let me know. I will close the push request when Im back home.

adrianschmutzler commented:
Okay, thank you for your work anyway. :-)

adrianschmutzler commented:
@wifiseguro 
Currently, this device calculates WAN mac address by just adding +1 to eth0 address.
However, can you check whether the WAN mac address is located on flash in <&factory 0xe006>or another location close to that?

wifiseguro commented:
> @wifiseguro
> Currently, this device calculates WAN mac address by just adding +1 to eth0 address.
> However, can you check whether the WAN mac address is located on flash in <&factory 0xe006>or another location close to that?
@adrianschmutzler 
I have no idea how to check that, can you please help me know?

adrianschmutzler commented:
Either do a hexdump of the whole factory partition or try the following:

. /etc/functions.sh
. /etc/functions/system.sh
mtd_get_mac_binary factory 0xe006

wifiseguro commented:
> Either do a hexdump of the whole factory partition or try the following:
> 
> . /etc/functions.sh
> . /etc/functions/system.sh
> mtd_get_mac_binary factory 0xe006

@adrianschmutzler 
Here is what I got when dumped factory partition:
000e000 a378 5c51 8e84 a378 5c51 8f84 ffff ffff
I dont know which exact byte is 0xe006 in that dump, but you can find two mac addresses in the bytes (in reverse order though)
a378 5c51 8e84
a378 5c51 8f84

Hope that helps.

adrianschmutzler commented:
Thanks, that's more than interesting. So, please run the following, which should return four "MAC addresses":

```
. /lib/functions.sh
. /lib/functions/system.sh
mtd_get_mac_binary factory 0x4
mtd_get_mac_binary factory 0x8004
mtd_get_mac_binary factory 0xe000
mtd_get_mac_binary factory 0xe006
```

(Note that I got the includes wrong in the previous post, it has to be lib instead of etc ...)

wifiseguro commented:
> Thanks, that's more than interesting. So, please run the following, which should return four "MAC addresses":
> 
> ```
> . /lib/functions.sh
> . /lib/functions/system.sh
> mtd_get_mac_binary factory 0x4
> mtd_get_mac_binary factory 0x8004
> mtd_get_mac_binary factory 0xe000
> mtd_get_mac_binary factory 0xe006
> ```
> 
> (Note that I got the includes wrong in the previous post, it has to be lib instead of etc ...)

@adrianschmutzler 
I get these values for the above:
78:a3:51:5c:84:8c
78:a3:51:5c:84:8d
78:a3:51:5c:84:8e
78:a3:51:5c:84:8f

adrianschmutzler commented:
Okay, thank you very much.
So 0xe006 is indeed holding the WAN mac address, I've added this to my PR #2319

wifiseguro commented:
> For reference:
> [6d91ddf](https://github.com/openwrt/openwrt/commit/6d91ddf5175d2eac3c4bc4a404cc0f5dd44cf25b)
> [a2c19f1](https://github.com/openwrt/openwrt/commit/a2c19f1d2f658367e6d62a6bdcfc72f12f23e43e)
@adrianschmutzler 
Would it be possible to pull that above two changes to 18.04 or 19.07 in order to have support for existing mt7621 device models with less than 512 of RAM?

adrianschmutzler commented:
> Would it be possible to pull that above two changes to 18.04 or 19.07 in order to have support for existing mt7621 device models with less than 512 of RAM?

Well, essentially this is not a fix, but a feature, so it is no typical candidate for a backport.
Despite, it should be easy for you to backport it yourself if it is only for this single device.
Same would have been true for device support PR BTW, we typically do not backport those ...

While I think this should apply to 18.06, I'm not so sure about 19.07 ATM.
The latter is not even in RC state, and I think adding the patches there would be a benefit worth the effort (also for supporting devices "automatically" in similar cases as the device discussed here).

However, the memory-autodetection patches have been added after the device rename patches in ramips. So, it would not be a trivial backport, as DTS names will be all different in 19.07. Fortunately, the second patch is just deleting ALL memory nodes (independent of the subtarget), so it should be relatively easy to script the same for 19.07 ramips DTSes.

@981213 What do you think about that?

Is there anything that I overlooked which blocks the backport of your patch (ramips: mt7621: add support for memory detection) to 19.07?

I think the amount of work is small enough and this would bring a valuable addition to the still-to-be-released stable branch.

If you do not want to backport the patches yourself, I would be willing to do it.

wifiseguro commented:
@adrianschmutzler 
Thank you very much for the explanations.I am willing to have that patches ported, if you think I can help please let me know.

981213 commented:
@adrianschmutzler 
I was told that upstream maintainers may not be happy with the "drop all memory nodes" approach, and I've just sent a mail to device-tree mailing list about this problem.

>I think the amount of work is small enough and this would bring a valuable addition to the still-to-be-released stable branch.
> 
> If you do not want to backport the patches yourself, I would be willing to do it.

I don't have much time for it recently. I think it's fine to backport memory detection and drop memory node for this single device only. As for other device-trees, I'm waiting for a reply upstream.

adrianschmutzler commented:
@981213 Thanks for the reply.
So, having a memory node won't collide with auto-detection for the other devices if we drop only for this device?

981213 commented:
@adrianschmutzler 

> So, having a memory node won't collide with auto-detection for the other devices if we drop only for this device?

Current code logic is this: If there's a memory node, use it. Otherwise, fallback to memory size detection. So no collision will happen.

adrianschmutzler commented:
Backport found in #2476
